nixos/tomcat: allow verbatim config override

polynomial opened the pull request:
###### Motivation for this change

I am attempting to use NixOS to replace several Amazon Linux Tomcat systems. The current tomcat module in NixOS doesn't allow for configuration overrides similar to other modules (see: nginx https://github.com/NixOS/nixpkgs/blob/release-17.09/nixos/modules/services/web-servers/nginx/default.nix#L314-L322 ).  The other two additions are patterns used in many enterprise tomcat deployments that:
* allow you to create log directories such as 'info/' and 'error/' in the logs directory (logback/slf4j expects the configuration/deployment system to create these directories if your logback.xml uses it, it won't create them for you)
* pass environment variables to tomcat which can be used to set certain configuration inside your applications

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

polynomial committed:
Tomcat Module:
* add serverXml verbatim override
* add environment file
* add log directory creation

Mic92 committed:
nixos/tomcat: add types + proper systemd integration

Mic92 commented:
Wow, there was quite a lot of old stuff in that module. I have added some stuff that might get in conflict with yours and needs to be discussed.

Mic92 commented on nixos/modules/services/web-servers/tomcat.nix:
```diff
@@ -36,42 +32,76 @@ in
+
+      environment = mkOption {
+        default = null;
+        example = "/etc/nixos/tomcat-environment";
```
A file of environment variables is not the nicest interface. Instead it should be a list of strings:

```
extraEnvironment = mkOption {
   type = types.listOf types.str;
   # ...
};
# ...
systemd.servies.tomcat.serviceConfig = {
  Environment = [
    # ...
   ] ++ extraEnvironment;
};
```

if somebody needs a file, it is always possible to set this manually outside the module: `systemd.servies.tomcat.serviceConfig.EnvironmentFile`
Or use nix to split a file in lines..

Mic92 commented on nixos/modules/services/web-servers/tomcat.nix:
```diff
@@ -275,94 +292,80 @@ in
+          "JAVA_HOME=${cfg.jdk}"
+          "CATALINA_OPTS=${cfg.catalinaOpts}"
+          "ERRFILE=SYSLOG"
+          "OUTFILE=SYSLOG"
```
It would be nice to log to journald by default instead of log files, 
but I also see why people want to have log files per vhost. 
Access log is set in the module a the moment. So maybe that should be disabled unless requested.

Mic92 replied:
Changing the logging scheme, should be probably guarded by `stateVersion`

Mic92 replied:
I think I will revert logging for the time being and this can addressed in future pull requests.

Mic92 commented:
Eval error is related to https://github.com/NixOS/nixpkgs/commit/7142440d75f8a55daf830ccf7628d4f739d124c5

grahamc commented:
@GrahamcOfBorg eval

danbst commented on nixos/modules/services/web-servers/tomcat.nix:
```diff
@@ -36,42 +32,76 @@ in
+        default = "";
+        description = "
+          Verbatim server.xml configuration.
+          This is mutualyl exclusive with the virtualHosts options.
```
`mutualyl` typo

Mic92 commented on nixos/modules/services/web-servers/tomcat.nix:
```diff
@@ -36,42 +32,76 @@ in
       };
 
       javaOpts = mkOption {
+        type = types.str;
```
We could accept both actually.

polynomial committed:
Updated for review comments:

* https://github.com/NixOS/nixpkgs/pull/35443#pullrequestreview-100238652
* https://github.com/NixOS/nixpkgs/pull/35443#pullrequestreview-100236904
* https://github.com/NixOS/nixpkgs/pull/35443#pullrequestreview-99112567

polynomial committed:
Remove environment setting, use list of strings instead

polynomial committed:
Change extraEnvironment, fix example typo

polynomial commented:
Thank you so much for the feedback, and thank you @Mic92 for the additions. I think I've addressed most of the comments. I didn't do anything about https://github.com/NixOS/nixpkgs/pull/35443#pullrequestreview-99112646 
https://github.com/NixOS/nixpkgs/pull/35443#pullrequestreview-100236616
Would love more direction about which specific course we want to take, and/or if we could future those as open issues and merge this as is.

Mic92 commented:
@polynomial Could you just change these two things?

- change `javaOpts` and `catalinaOpts` to type `type = types.either (types.listOf types.str) types.str;` and put a `toStrings` where they are used to convert them?
- remove "ERRFILE=SYSLOG" and "OUTFILE=SYSLOG" again and make sure log files work as advertised

I think other stuff should be not solved in the scope of this pull request.

polynomial committed:
Updated per review comments

* add javaOpts to systemd service definition
* remove syslog defaults

polynomial commented:
@Mic92 Done! I tested locally and the javaOpts seemed to have been entirely broken! I think it is fixed now and I tested it locally (and the log change too). Let me know if there is anything else, have a great day. (:

Mic92 commented:
rebased and pushed in 196e21a1606ee024f5228d2f83d72610aa495ed0

pvgoran commented:
Any reason why this wasn't merged into the stable release? I want to make further changes to the Tomcat module, and it's difficult to do when the differences between stable and master are that big.

Mic92 commented:
We are conservative about back porting to not break the stable release. And this was a rather big change.

pvgoran commented:
@Mic92 I see. But when and how will the back porting eventually happen? What is the procedure? The change is already rather old.

pvgoran commented:
@Mic92 @polynomial What's the deal with the type of `virtualHosts` in this PR? Why it its `name` attribute `types.listOf types.str` and not just `types.str`? And its `webapps` attribute is not described at all!

danbst commented:
@pvgoran you may switch to `nixos-unstable`. Other than that, you can replace `tomcat.nix` NixOS module with one from `nixos-unstable` (see https://nixos.org/nixos/manual/#sec-replace-modules for more details).

> But when and how will the back porting eventually happen? What is the procedure? The change is already rather old.
No backporting. It will be available in stable NixOS 18.09, which will be ready in October (2 months left to release)

> What's the deal with the type of virtualHosts in this PR? Why is its name attribute types.listOf types.str and not just types.str? And its webapps attribute is not described at all!
IMO this is stuff from old author, @svanderburg. Ping him if you want some clarifications. Though `webapps` isn't described, I did use it for deploying tomcat webapps.

pvgoran commented:
>  Other than that, you can replace tomcat.nix NixOS module with one from nixos-unstable (see https://nixos.org/nixos/manual/#sec-replace-modules for more details).

Thanks, it will help.

> No backporting. It will be available in stable NixOS 18.09, which will be ready in October (2 months left to release)

It's a bit sad, but I understand that the proper release process is better suited for rolling out significant changes. By the way, is there any kid of policy document that describes what should be backported?

> IMO this is stuff from old author, @svanderburg. Ping him if you want some clarifications. Though `webapps` isn't described, I did use it for deploying tomcat webapps.

The type for `virtualHosts` was introduced in this PR, in the following commit: https://github.com/NixOS/nixpkgs/pull/35443/commits/472f16d7fbcd19211f6d0e88a773b4445a49b46e#diff-1ca47ff430c580c1c14cf86b7bd32028R117, apparently authored by @Mic92. So @svanderburg doesn't seem to have anything to do with it.

Now, the thing is, the type specification is indeed wrong. My configuration that works on release-18.03 breaks when I build it with a module from nixos-unstable. First it complains:

```
error: The option `services.tomcat.virtualHosts.[definition 1-entry 1].webapps' defined in `<somefile>.nix' does not exist.
```

If I remove the `webapps` attribute, it complains about `name`:

```
error: The option value `services.tomcat.virtualHosts.[definition 1-entry 1].name' in `<somefile>.nix' is not of type `list of strings'.
```

Finally, if I change `name` into a list of strings, it fails with:

```
error: cannot coerce a list to a string, at /nix/var/nix/profiles/per-user/root/channels/nixos-unstable/nixos/modules/services/web-servers/tomcat.nix:218:107
```

I'll prepare a correction. It would be a separate PR, since this one is long merged, right?

danbst commented:
> The type for virtualHosts was introduced in this PR, in the following commit:
Ah, right! You were talking about type, not an option.

Indeed, if you experience backwards incompatible changes, you should make issues/PRs. But please, show exceprt of original config, so issue is reproducible.

> By the way, is there any kid of policy document that describes what should be backported?
There probably is one. The rule of thumb is "backport only security patches" and internal bug fixes, no package/modules featureful updates. In rare cases, maintainers way lobby some non-security stuff when they have time and need.

pvgoran commented:
Created https://github.com/NixOS/nixpkgs/issues/44361.
